types/fullcalendar__vue/index.d.ts

```diff
@@ -0,0 +1,193 @@
+    eventOverlap: object;
+    eventAllow: object;
+    eventClassName: object;
+    eventClassNames: object;
+    eventBackgroundColor: object;
+    eventBorderColor: object;
+    eventTextColor: object;
+    eventColor: object;
+    events: object;
+    eventSources: object;
+    allDayDefault: object;
+    startParam: object;
+    endParam: object;
+    lazyFetching: object;
+    nextDayThreshold: object;
+    eventOrder: object;
+    rerenderDelay: object;
+    dragRevertDuration: object;
+    dragScroll: object;
+    longPressDelay: object;
+    eventLongPressDelay: object;
+    droppable: object;
+    dropAccept: object;
+    eventDataTransform: object;
+    allDayMaintainDuration: object;
+    eventResizableFromStart: object;
+    timeGridEventMinHeight: object;
+    allDayHtml: object;
+    eventDragMinDistance: object;
+    eventResourceEditable: object;
+    eventSourceFailure: object;
+    eventSourceSuccess: object;
+    forceEventDuration: object;
+    progressiveEventRendering: object;
+    selectLongPressDelay: object;
+    selectMinDistance: object;
+    timeZoneParam: object;
+    titleRangeSeparator: object;
+    // compound OptionsInput...
+    buttonText: object;
+    views: object;
+    plugins: object;
+    // scheduler...
+    schedulerLicenseKey: object;
+    resources: object;
+    resourceLabelText: object;
+    resourceOrder: object;
+    filterResourcesWithEvents: object;
+    resourceText: object;
+    resourceGroupField: object;
+    resourceGroupText: object;
+    resourceAreaWidth: object;
+    resourceColumns: object;
+    resourcesInitiallyExpanded: object;
+    slotWidth: object;
+    datesAboveResources: object;
+    googleCalendarApiKey: object;
+    refetchResourcesOnNavigate: object;
+    // used to be emissions but are now props...
+    datesRender: object;
+    datesDestroy: object;
+    dayRender: object;
+    eventRender: object;
+    eventDestroy: object;
+    viewSkeletonRender: object;
+    viewSkeletonDestroy: object;
+    resourceRender: object;
+}
+
+interface PropIsDeep {
+    header: boolean;
+    footer: boolean;
+    events: boolean;
+    eventSources: boolean;
+    resources: boolean;
+}
+
+interface EmissionUseProp {
+    datesRender: boolean;
+    datesDestroy: boolean;
+    dayRender: boolean;
+    eventRender: boolean;
+    eventDestroy: boolean;
+    viewSkeletonRender: boolean;
+    viewSkeletonDestroy: boolean;
+    resourceRender: boolean;
+}
+
+interface CalendarComponent extends Vue {
+    buildCalendarOptions: () => void;
+    recordDirtyOption: (optionName: any, newVal: any) => void;
+    renderDirty: () => void;
+    getApi: () => any;
+}
+
+export type FullCalendarComponent = CalendarComponent;
+export const PROP_DEFS: PropsDef;
+export const PROP_IS_DEEP: PropIsDeep;
+export const EMISSION_NAMES: string[];
+export const EMISSION_USE_PROP: EmissionUseProp;
+
+export function intall(Vue: () => Vue): void;
+
+export const FullCalendar: VueConstructor;
+export default FullCalendar;
```

Comment:
I may just be reading this incorrectly, but it looks like the default export of this module is a re-export of a symbol from Vue?  That seems odd.

Reply:
Hi thanks for reviewing this soo quickly. I am new to writing type defs, first attempt so guidance would be appreciated. I orignally tried to emulate what I saw in this package https://github.com/DefinitelyTyped/DefinitelyTyped/blob/aa5c66c087896ff55568b76ae49f77321ba0123a/types/vue-datetime/index.d.ts but looking closer I think maybe I should be emulating this package instead https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5d5716a9c3c942ec762094fab566729002365a5e/types/vue2-datepicker/index.d.ts

Originally I had created an interface that extended the vue constructor but the linter complained there shouldn't be an empty interface.

Reply:
So I have made quite a big change, I was originally following a pattern for components built for vue v1 I think and found vue2 packages.  I have rewritten this following the pattern for https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5d5716a9c3c942ec762094fab566729002365a5e/types/vue2-datepicker/index.d.ts

So now I am exporting the following
```
declare const FullCalendar: Component<any, any, any, CalendarProps>;

export function install(Vue: () => any): void;
export type FullCalendarComponent = CalendarComponent;
export default FullCalendar;
``` 

CalendarProps is an interface for the Vue props I declared in an options.d.ts file and imported it into index.d.ts so as to not make it too cluttered but I can move it all into 1 file.

This is quite a big change so should I make a new PR or are you happy reviewing it as part of this PR?

Reply:
Sorry for the delay - I've been looking at other things.  This is a new definition, so I'm happy to review everything at once in this PR.

Reply:
I guess I'm still surprised `FullCalendar` and `FullCalendarComponent` aren't related, but I don't know anything about the underlying library and it may well be correct.

Reply:
This may very well be my missunderstanding of how to create type defs so forgive, I'll try to explain the situation I am tryong to solve. 

The most basic part is the FullCalendar declaration 
`declare const FullCalendar: Component<any, any, any, CalendarProps>;` , this allows the user to load the component into another component for use in the template or in a SFC.

e.g:
```
components: {
        'full-calendar': FullCalendar,
},
template: `
    <full-calendar defaultView="dayGridMonth" :plugins="calendarPlugins" ref="fullCalendar"/>
    `,
```

The FullCalendarComponent type is to easily deal with a querk in Vue js when it comes to refrencing a child components methods using `$refs`. It is based on this issue https://github.com/vuejs/vue-class-component/issues/94

To access the core FullCalendar library api one has to reference the component as such [according to the official docs](https://fullcalendar.io/docs/vue)
```
let calendarApi = this.$refs.fullCalendar.getApi()
calendarApi.next() 
```

Doing so with typescript makes the compiler kick up a stink so the offical docs are wrong. 

There is 2 ways of resolving this 
```
// Not sure this is the nicest way 
let calendarApi = (this.$refs.fullCalendar as any).getApi();

// or
let calendarApi = (this.$refs.fullCalendar as FullCalendarComponent).getApi();

// Or with the class based api
class YourComponent extends Vue {
  $refs!: {
    fullCalendar : FullCalendarComponent
  }
}
```

I figured having a type which could be imported and which reflects the api was nicer than having to either define your own interface everytime or declaring it as any. I may be wrong and feedback is welcome.

Reply:
It seems like something's gone wrong somewhere along the line if you need a type assertion to access `getApi`, but the bug you linked suggests that it's in the project itself.  Unfortunately, I don't know enough about this library, or Vue in general, to be able to suggest a better alternative.  I will, however, suggest that you just `export interface FullCalendarComponent extends Vue`, so you don't have a redundant type name floating around.  Otherwise, it seems reasonable and it's easy enough to change later.

Reply:
Hi thanks for the guidance. I have updated the defs so the interface FullCalendarComponent is exported rather than just the type alias.